Cosmetics Editor entry for Ivan the Fairy

skyyoshi86 opened the pull request:
![image](https://github.com/HarbourMasters/Shipwright/assets/105878730/0bc9f908-1039-4202-b537-f9be8ed93c6d)


Added Cosmetics Editor entry to change the color of Ivan the Fairy

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1138407827.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1138407829.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1138407830.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1138407832.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1138407833.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1138407834.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1138407835.zip)
<!--- section:artifacts:end -->

skyyoshi86 committed:
First commit

jbodner09 reviewed (changes requested):
The changes to the CMakeSettings and MODDING.md seem unrelated to Ivan and shouldn't be included in this PR.

skyyoshi86 committed:
Reverted MODDING.md

Reverted a correction to MODDING.md (case-sensitive broken link?) as this is not a change related to this PR

skyyoshi86 committed:
Delete CMakeSettings.json

Deleted CMakeSettings.json as this is not a change related to this PR.

skyyoshi86 commented:
> The changes to the CMakeSettings and MODDING.md seem unrelated to Ivan and shouldn't be included in this PR.

TY for Review! I'm learning. CMakeSettings.json file was generated/changed automatically, and now I don't know if I need to delete it. I now deleted it. Was it correct to delete it?

jbodner09 commented:
Yeah, it was most likely auto-generated by your IDE, but shouldn't have been.  If other people run into this, it would be easy enough to add it to the git ignore list, but at least on my machine, the file doesn't exist, so I'm curious what your setup is.

skyyoshi86 commented:
> Yeah, it was most likely auto-generated by your IDE, but shouldn't have been. If other people run into this, it would be easy enough to add it to the git ignore list, but at least on my machine, the file doesn't exist, so I'm curious what your setup is.

TY! I'm on Windows 10 using Visual Studio 2022. As I kept getting an "could not clone vcpkg" when trying to compile, I was told to use "& 'C:\Program Files\CMake\bin\cmake' -S . -B "build/x64" -G "Visual Studio 17 2022" -T v142 -A x64 -DVCPKG_ROOT="build/x64/vcpkg" -DBUILD_REMOTE_CONTROL=1" command when compiling. I was told that "It's trying to update the vcpkg that came installed with Visual Studio, rather than making a separate vcpkg clone just in the project."

skyyoshi86 commented:
Am I missing a step for this PR?

briaguya0 commented on docs/MODDING.md:
````diff
@@ -188,4 +188,4 @@ Assuming all went well, you can now push your changes to your fork with the foll
 git push origin <YOUR BRANCH NAME>
-```
\ No newline at end of file
+```
````
anyone know what this change is? it doesn't appear to be newline/not

jbodner09 replied:
Well it is, but github apparently displays it differently for markdown files than it does for code files.  I checked out this branch and opened the file up in a hex editor just to confirm:  
![image](https://github.com/HarbourMasters/Shipwright/assets/30329717/571eff9d-ba99-4b43-9e40-d0fc9cf060e8)

jbodner09 replied:
Since the other markdown files end in newlines, it should be fine to leave it.

Archez commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -850,14 +862,28 @@ void EnPartner_Draw(Actor* thisx, PlayState* play) {
     gDPPipeSync(dListHead++);
-    gDPSetPrimColor(dListHead++, 0, 0x01, (u8)this->innerColor.r, (u8)this->innerColor.g, (u8)this->innerColor.b,
-                    (u8)(this->innerColor.a * alphaScale));
+    if (CVarGetInteger("gCosmetics.Ivan_IdlePrimary.Changed", 0)) {
+        Color_RGB8 ivanColor1 = CVarGetColor24("gCosmetics.Ivan_IdlePrimary.Value", (Color_RGB8){ 255, 255, 255 });
+        gDPSetPrimColor(dListHead++, 0, 0x01, (u8)ivanColor1.r, (u8)ivanColor1.g, (u8)ivanColor1.b,
+                        (u8)(this->innerColor.a * alphaScale));
```
Since the body and particle colors are the same (both default green or the cosmetic color), instead of doing this lookup twice in each draw func, you can instead do it once in the `EnPartner_Update` func, and have it set `this->innerColor` and `this->outerColor` to the right color, then the two draw funcs can be reverted to use those values again.

skyyoshi86 replied:
> Since the body and particle colors are the same (both default green or the cosmetic color), instead of doing this lookup twice in each draw func, you can instead do it once in the `EnPartner_Update` func, and have it set `this->innerColor` and `this->outerColor` to the right color, then the two draw funcs can be reverted to use those values again.

![image](https://github.com/HarbourMasters/Shipwright/assets/105878730/05e10f40-4897-4beb-af56-f7a81bcc7aae)


TY for your help and feedback :D

1. How can I get the colors from Cosmetics Editor with a function similar to `CVarGetColor24()` that returns a `Color_RGBAf` value instead of returning a `Color_RGB8` value?

-or-

2. Should I instead modify the declarations of `this->innerColor` and `this->outerColor` to be of type `Color_RGB8` so I can assign the Cosmetic Editor colors via the `CVarGetColor24()` function?

EDIT:

3. Should I keep using my newly created variables `ivanColor1` and `ivanColor2` of type `Color_RGB8` as an alternative to assign the colors if they have been changed in the Cosmetics Editor?

Thank you in advance :)

Archez replied:
You can assign the individual color properties one by one, which will handle the casting from int to float.

```c
Color_RGB8 idlePrim = CVarGetColor24(...);
this->innerColor.r = idlePrim.r;
this->innerColor.g = idlePrim.g;
...
```

Or if you are daring, you can change the `Color_RGBAf` in the EnPartner struct to be `Color_RGBA8`. Either way is fine.

skyyoshi86 committed:
checks for cosmetic changes ONCE for both body and sparles

Please review carefully. I made these changes blindly as I figure out why I can't compile it in my machine anymore. Sorry!

skyyoshi86 committed:
correctly updates OUTER color when it's changed in Cosmetics Editor

Archez reviewed (approved):
Awesome, looks good now! 🙂

garrettjoecox commented on soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp:
```diff
@@ -1619,13 +1624,15 @@ void RandomizeColor(CosmeticOption& cosmeticOption) {
         CopyMultipliedColor(cosmeticOption, cosmeticOption, 4.0f);
     } else if (cosmeticOption.label == "Idle Primary") {
-        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_IdleSecondary"), 0.5f);
+            CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_IdleSecondary"), 0.5f);
```
```suggestion
        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_IdleSecondary"), 0.5f);
```

garrettjoecox committed:
Update soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp

garrettjoecox commented on soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp:
```diff
@@ -1621,11 +1626,13 @@ void RandomizeColor(CosmeticOption& cosmeticOption) {
         CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_IdleSecondary"), 0.5f);
     } else if (cosmeticOption.label == "Enemy Primary") {
-        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_EnemySecondary"), 0.5f);
+            CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_EnemySecondary"), 0.5f);
```
```suggestion
        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_EnemySecondary"), 0.5f);
```

garrettjoecox committed:
Update soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp

garrettjoecox commented on soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp:
```diff
@@ -1623,9 +1628,11 @@ void RandomizeColor(CosmeticOption& cosmeticOption) {
         CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_EnemySecondary"), 0.5f);
     } else if (cosmeticOption.label == "NPC Primary") {
-        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_NPCSecondary"), 1.0f);
+            CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_NPCSecondary"), 1.0f);
```
```suggestion
        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_NPCSecondary"), 1.0f);
```

garrettjoecox committed:
Update soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp

garrettjoecox commented on soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp:
```diff
@@ -1625,7 +1630,9 @@ void RandomizeColor(CosmeticOption& cosmeticOption) {
         CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_NPCSecondary"), 1.0f);
     } else if (cosmeticOption.label == "Props Primary") {
-        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_PropsSecondary"), 1.0f);
+            CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_PropsSecondary"), 1.0f);
```
```suggestion
        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_PropsSecondary"), 1.0f);
```

garrettjoecox committed:
Update soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp

garrettjoecox commented on soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp:
```diff
@@ -1626,6 +1631,8 @@ void RandomizeColor(CosmeticOption& cosmeticOption) {
     } else if (cosmeticOption.label == "Props Primary") {
         CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Navi_PropsSecondary"), 1.0f);
+    } else if (cosmeticOption.label == "Ivan Idle Primary") {
+            CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Ivan_IdleSecondary"), 0.5f);
```
```suggestion
        CopyMultipliedColor(cosmeticOption, cosmeticOptions.at("Ivan_IdleSecondary"), 0.5f);
```

garrettjoecox committed:
Update soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -790,7 +812,7 @@ void DrawOrb(Actor* thisx, PlayState* play, u8 color) {
 
     gDPSetPrimColor(POLY_XLU_DISP++, 0, 0x80, 255, 255, 255, 255);
-
+     
```
```suggestion

```

garrettjoecox committed:
Update soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -155,7 +155,7 @@ void EnPartner_SpawnSparkles(EnPartner* this, PlayState* play, s32 sparkleLife)
     primColor.g = this->innerColor.g;
     primColor.b = this->innerColor.b;
-
+ 
```
```suggestion

```

garrettjoecox committed:
Update soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c

briaguya0 commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -790,7 +812,6 @@ void DrawOrb(Actor* thisx, PlayState* play, u8 color) {
 
     gDPSetPrimColor(POLY_XLU_DISP++, 0, 0x80, 255, 255, 255, 255);
-
     switch (color) {
```
```suggestion

    switch (color) {
```

garrettjoecox committed:
Update soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c

Co-authored-by: briaguya <70942617+briaguya-ai@users.noreply.github.com>

briaguya0 commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -155,7 +155,6 @@ void EnPartner_SpawnSparkles(EnPartner* this, PlayState* play, s32 sparkleLife)
     primColor.g = this->innerColor.g;
     primColor.b = this->innerColor.b;
-
     envColor.r = this->outerColor.r;
```
```suggestion

    envColor.r = this->outerColor.r;
```

garrettjoecox committed:
Update soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c

Co-authored-by: briaguya <70942617+briaguya-ai@users.noreply.github.com>

briaguya0 commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -851,13 +872,15 @@ void EnPartner_Draw(Actor* thisx, PlayState* play) {
     gDPPipeSync(dListHead++);
     gDPSetPrimColor(dListHead++, 0, 0x01, (u8)this->innerColor.r, (u8)this->innerColor.g, (u8)this->innerColor.b,
-                    (u8)(this->innerColor.a * alphaScale));
+                        (u8)(this->innerColor.a * alphaScale));
+    
```
```suggestion
                    (u8)(this->innerColor.a * alphaScale));
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -851,13 +872,15 @@ void EnPartner_Draw(Actor* thisx, PlayState* play) {
     gSPEndDisplayList(dListHead++);
     gDPSetEnvColor(POLY_XLU_DISP++, (u8)this->outerColor.r, (u8)this->outerColor.g, (u8)this->outerColor.b,
-                   (u8)(envAlpha * alphaScale));
+                      (u8)(envAlpha * alphaScale));  
+
```
```suggestion
                  (u8)(envAlpha * alphaScale));
```

garrettjoecox committed:
Update soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c

Co-authored-by: briaguya <70942617+briaguya-ai@users.noreply.github.com>

garrettjoecox committed:
Update soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c

Co-authored-by: briaguya <70942617+briaguya-ai@users.noreply.github.com>

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c:
```diff
@@ -857,7 +879,7 @@ void EnPartner_Draw(Actor* thisx, PlayState* play) {
     gSPEndDisplayList(dListHead++);
     gDPSetEnvColor(POLY_XLU_DISP++, (u8)this->outerColor.r, (u8)this->outerColor.g, (u8)this->outerColor.b,
-                   (u8)(envAlpha * alphaScale));
+                  (u8)(envAlpha * alphaScale));
```
```suggestion
                   (u8)(envAlpha * alphaScale));
```

garrettjoecox committed:
Update soh/src/overlays/actors/ovl_En_Partner/z_en_partner.c
